### PR TITLE
Migration to Python3

### PR DESCRIPTION
**Pull request by mpgn**
<del>**Don't merge anything yet, this PR is just here to keep you informate of the progress**</del>
---

**Progression**

- [x] --shares
- [x]  --sessions 
- [x]  --disks 
- [x]  --loggedon-users 
- [x]  --users
- [x]  --rid-brute
- [x]  --groups
- [x]  --local-groups 
- [x]  --pass-pol (local)
- [x] --pass-pol (domain)
- [x]  -x whoami
- [x] -X '$PSVersionTable' 
- [x] --sam
- [x] -H ntlm_hash
- [x] -u user.txt -p password.txt
- [x] --gen-relay-list
- [x] -d domain.local
- [x] --local-auth
- [x] --lsa
- [x] --ntds
- [x] <del>--ntds-history</del> option already deprecated in CME v4 python2
- [x] <del>--ntds-pwdLastSe</del> option already deprecated in CME v4 python2
- [x] --wmi 
- [ ] --exec-method smbexec -> not working but I got same result on CME v4 python2
- [x] --exec-method atexec
- [x] --exec-method wmiexec
- [x] --no-output
- [x] cmedb
- [x] --spider -> error from impacket lib, issue <del>open</del> **fixed** https://github.com/SecureAuthCorp/impacket/issues/691

---

- [x] module Mimikatz
- [x] module Bloodhound

**Modules** modules based on Empire project should be removed since the Empire project is dead and dev in python2. I didn't test them, hope it makes sense.

---
- [x] migrate pywerview to python3 https://github.com/mpgn/pywerview
- [x] pull request to pywerview repo (waiting for the merge https://github.com/the-useless-one/pywerview/pull/34)
- [ ] waiting for the merge of pywerview in the meantime we use my fork of pywerview 
- [x] migrate Invoke-vnc to python3 https://github.com/klsecservices/Invoke-Vnc/pull/2
- [ ] waiting for the merge of Invoke-vnc
---

<del>So far, I migrate the wmiexec.py file and lot of other stuff just to test how hard it will be to migrate :) </del>

I will continue the work over this week.

image

How to use this PR ?

- with pipenv
```
git clone --recursive https://github.com/mpgn/CrackMapExec && cd CrackMapExec
git checkout python3
git submodule update --recursive
pipenv --python 3.7 install
pipenv shell
python setup.py install
cme
```

On latest Kali
```
git clone --recursive https://github.com/mpgn/CrackMapExec && cd CrackMapExec
git checkout python3
git submodule update --recursive
python3 setup.py install
cme
```

--- 

Couple of things @byt3bl33d3r  wants to take care of before merging this:
- [ ] Get rid of all of the submodules (This also would mean getting Pywerview on Pypi)
- [ ] Replace Gevent with Asyncio
- [ ] Package CME as a Zipapp with Shiv so people don't have to worry about installing cme in a virtualenv anymore

### CODE REVIEW

File: cme/protocols/smb/wmiexec.py [hunk 1]
```diff
@@ -8,121 +8,122 @@
...
+			self.__outputBuffer = ''
+
+	def output_callback(self, data):
+		self.__outputBuffer += data.decode("utf-8") 
+
+	def execute_handler(self, data):
+		if self.__retOutput:
+			try:
+				logging.debug('Executing remote')
+				self.execute_remote(data)
```
Comment (mpgn) [inline cme/protocols/smb/wmiexec.py]: for the purpose of the test, I use `execute_remote` to get result of my command on Win10.

**Comment by awsmhacks** [comment]
yo @mpgn , i've got quite a bit of this figured out (on the smb side) if you want to take a look. https://github.com/awsmhacks/CrackMapExtreme   
all the powerview functionality was ported straight into the smb file.  Check the enum functions for examples.
https://github.com/awsmhacks/CrackMapExtreme/blob/master/cmx/protocols/smb.py#L1365

**Comment by mpgn** [comment]
thx @awsmhacks I will look at it :)

So far I didn't encourted much trouble for the migration as CME is well coded. Finger cross :100:

**Comment by byt3bl33d3r** [comment]
Thanks for the help @mpgn. This is awesome. Can't wait to review it :)

**Comment by byt3bl33d3r** [comment]
Hey @mpgn, i switched the target branch for this commit to `python3` so that we can test changes before merging them to master.

**Comment by mpgn** [comment]
Just updated pywerview, now option --groups and --users is working like a charm on CME :)

**Comment by mpgn** [comment]
@byt3bl33d3r I open the pull request so people can test this branch :+1:

**Comment by byt3bl33d3r** [comment]
Dope! Couple of things I want to take care of before merging this:
- [] Get rid of all of the submodules (This also would mean getting Pywerview on Pypi)
- [] Replace Gevent with Asyncio
- [] Package CME as a Zipapp with Shiv so people don't have to worry about installing cme in a virtualenv anymore

**Comment by byt3bl33d3r** [comment]
@mpgn have you tested `cmedb` with the new changes by any chance?

**Comment by mpgn** [comment]
@byt3bl33d3r `cmedb` next, added to the progression checklist :)

**Comment by mpgn** [comment]
image

**Comment by mpgn** [comment]
@byt3bl33d3r quick question, why you want to replace Gevent by Asyncio ?

**Comment by byt3bl33d3r** [comment]
1 less dependency :) plus IMHO asyncio would make things a lot simpler.

**Comment by mpgn** [comment]
I will look at this over the week end :)

**Comment by byt3bl33d3r** [comment]
@mpgn oh i'll take care of that part! (that's what i meant lol) you already did an insane amount

**Comment by mpgn** [comment]
Issue https://github.com/SecureAuthCorp/impacket/issues/691 has been resolved :+1:

**Comment by mpgn** [comment]
Hello @byt3bl33d3r, Happy New Year 💫

Any update now python2 is deprecated ?

**Comment by mpgn** [comment]
Hello again @byt3bl33d3r 

Maybe we could merge this to the python3 branch of CME without asyncio ?

**Comment by rhertzog** [comment]
FWIW, there's a maintained fork of Empire that has Python 3 support: https://github.com/BC-SECURITY/Empire (it's packaged in Kali)

**Comment by mpgn** [comment]
Module Mimikatz, Bloodhound etc should now be compatible with python3 ! :boom:

Capture
blood

**Comment by ValZapod** [comment]
This should also be fixed (when you apt-get install)
crackmapexec (4.0.1+git20200118-0kali1) …
/usr/lib/python3/dist-packages/cme/protocols/mssql.py:71: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if self.args.auth_type is 'windows':
/usr/lib/python3/dist-packages/cme/protocols/mssql.py:143: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if self.args.auth_type is 'windows':
/usr/lib/python3/dist-packages/cme/protocols/mssql.py:159: SyntaxWarning: "is" with a literal. Did you mean "=="?
  res = self.conn.login(None, username, password, domain, None, True if self.args.auth_type is 'windows' else False)
/usr/lib/python3/dist-packages/cme/protocols/mssql.py:173: SyntaxWarning: "is" with a literal. Did you mean "=="?
  out = u'{}{}:{} {}'.format('{}\\'.format(domain) if self.args.auth_type is 'windows' else '',
/usr/lib/python3/dist-packages/cme/protocols/smb.py:910: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if self.args.ntds is 'vss':

**Comment by mpgn** [comment]
Only one thirdparty remaining => pywerview

**Pull request merged**